cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -16,10 +16,213 @@
 
 #pragma once
 
+#include <cstdint>
 #include <cudf/strings/string_view.cuh>
+#include <hash/hash_constants.hpp>
+
+#include "cuda_runtime_api.h"
+#include "cudf/types.hpp"
+#include "driver_types.h"
+#include "vector_types.h"
 
 using hash_value_type = uint32_t;
 
+namespace cudf {
+namespace detail {
+
+/**
+ * @brief Core MD5 algorith implementation. Processes a single 512-bit chunk,
+ * updating the hash value so far. Does not zero out the buffer contents.
+ */
+void CUDA_DEVICE_CALLABLE md5_hash_step(md5_intermediate_data* hash_state,
+                                        md5_hash_constants_type const* hash_constants,
+                                        md5_shift_constants_type const* shift_constants)
+{
+  uint32_t A = hash_state->hash_value[0];
+  uint32_t B = hash_state->hash_value[1];
+  uint32_t C = hash_state->hash_value[2];
+  uint32_t D = hash_state->hash_value[3];
+
+  for (unsigned int j = 0; j < 64; j++) {
+    uint32_t F;
+    uint32_t g;
+    switch (j / 16) {
+      case 0:
+        F = (B & C) | ((~B) & D);
+        g = j;
+        break;
+      case 1:
+        F = (D & B) | ((~D) & C);
+        g = (5 * j + 1) % 16;
+        break;
+      case 2:
+        F = B ^ C ^ D;
+        g = (3 * j + 5) % 16;
+        break;
+      case 3:
+        F = C ^ (B | (~D));
+        g = (7 * j) % 16;
+        break;
+    }
+
+    uint32_t buffer_element_as_int;
+    thrust::copy_n(thrust::seq,
+                   hash_state->buffer + g * 4,
+                   4,
+                   reinterpret_cast<uint8_t*>(&buffer_element_as_int));
+    F = F + A + hash_constants[j] + buffer_element_as_int;
+    A = D;
+    D = C;
+    C = B;
+
+    uint32_t shift = shift_constants[((j / 16) * 4) + (j % 4)];
+    B              = B + ((F << shift) | (F >> (32 - shift)));
+  }
+
+  hash_state->hash_value[0] += A;
+  hash_state->hash_value[1] += B;
+  hash_state->hash_value[2] += C;
+  hash_state->hash_value[3] += D;
+
+  hash_state->buffer_length = 0;
+}
+
+template <typename Key>
+struct MD5Hash {
```

Comment:
Why is `MD5Hash` templated on `Key` in addition to the `operator()`s being templates? I would think the struct does not need to be a template and instead just make the `operator()` be a template. That would simplify your specializations.

Reply:
I was largely copying the structure of the murmur hash function, I'll change the operator to a template.

Reply:
Yes, but notice that `MurmurHas3_32` is a class template:
https://github.com/rapidsai/cudf/blob/855e735c07e14d11b8ef309d943e9a28c91e0698/cpp/include/cudf/detail/utilities/hash_functions.cuh#L32-L33

But the `operator()` is not: https://github.com/rapidsai/cudf/blob/855e735c07e14d11b8ef309d943e9a28c91e0698/cpp/include/cudf/detail/utilities/hash_functions.cuh#L76

In your case, both are templates, which just complicates the specializations.

Reply:
In order to fix the type dispatching error this was changed. The base operator is no longer templated https://github.com/rapidsai/cudf/pull/5438/files#diff-a6ce3f9a4f61a23dd6469473c7dbf15fR147

Reply:
Removed the template when I removed the md5_element_hasher